### PR TITLE
Worker enhancements

### LINKED ISSUE #9: Use Worker API to execute compilation
The Worker API is Gradle 4.1's new way to kick off separate processes during plugins.

### PR DESCRIPTION
**Pull request by ajoberstar**
The worker API is meant to be an efficiency gain in two ways: reusing warm/compatible JVMs, allowing for parallelizing  of tasks. The JVM reuse is good and bad for Clojure and some of our specific tasks. Starting a JVM is a cost that is nice to get rid of, starting the Clojure runtime even more so. But particularly for compilation, this can be problematic if classes that were loaded previously no longer exist. You might be expecting a NoClassDefFoundError, but get a successful compile.

To avoid this, I'm trying to detect whether an incompatible change has been made on the classpath, and restarting the Clojure runtime if it was. Leveraging [shimdandy](https://github.com/projectodd/shimdandy) to isolate the runtime and clean it up if we need a fresh one.

This also involved refactoring again into two different modules: the plugin itself, and a tools JAR that is used by the workers. It includes the compiler and test runner and any helper classes. Due to the way the workers use the Gradle classpath, it's helpful to keep these outside of the plugin so they can be added to the shim's classpath, but not the worker as a whole.

This is also restores the fail on reflection support, and adds tests to validate.

On your first build this is a little slower to use than the other method. But once the daemon and workers are running it speeds things up.

This PR resolves #9.

### CODE REVIEW

**Comment by ajoberstar** [comment]
This is now ready for review, when someone's got time. It's probably worth reviewing the commits independently, they're more logical at this point.

**Comment by ajoberstar** [comment]
Just updated the build to upload to Bintray, in addition to the plugin portal. This is needed to make the tools JAR accessible. I've put in the request to have it included in JCenter, as well.

If any of you have a Bintray account, let me know, and I'll get you added to the [gradle-clojure org](https://bintray.com/gradle-clojure) out there.

**Comment by jszakmeister** [comment]
I missed the fact that this was marked as ready for review... It looks like we have some conflicts now though.

**Comment by ajoberstar** [comment]
I should be able to take a look at the conflicts this evening.

**Comment by jszakmeister** [comment]
No rush.  It'll probably be the weekend before I can get a good look at it--my evenings this week are pretty full.

**Comment by ajoberstar** [comment]
Conflicts are fixed

File: modules/gradle-clojure-plugin/src/main/java/gradle_clojure/plugin/tasks/ClojureCompileOptions.java [hunk 1]
```diff
@@ -15,12 +15,13 @@
...
 package gradle_clojure.plugin.tasks;
 
+import java.io.Serializable;
 import java.util.Collections;
 import java.util.List;
 
 import org.gradle.api.Action;
 
-public final class ClojureCompileOptions {
+public final class ClojureCompileOptions implements Serializable {
```
Comment (jszakmeister) [inline modules/gradle-clojure-plugin/src/main/java/gradle_clojure/plugin/tasks/ClojureCompileOptions.java]: Do we need a `serialVersionUID` now?

File: modules/gradle-clojure-plugin/src/main/java/gradle_clojure/plugin/tasks/ReflectionWarnings.java [hunk 2]
```diff
@@ -15,7 +15,9 @@
  */
 package gradle_clojure.plugin.tasks;
 
-public final class ReflectionWarnings {
+import java.io.Serializable;
+
+public final class ReflectionWarnings implements Serializable {
```
Comment (jszakmeister) [inline modules/gradle-clojure-plugin/src/main/java/gradle_clojure/plugin/tasks/ReflectionWarnings.java]: Same here... do we need a `serialVersionUID`?

File: modules/gradle-clojure-tools/build.gradle [hunk 3]
```diff
@@ -7,6 +7,8 @@ plugins {
 dependencies {
   // gradle
   compileOnly gradleApi()
+  compileOnly 'org.clojure:clojure:1.8.0'
+  compileOnly 'org.projectodd.shimdandy:shimdandy-api:1.2.0'
```
Comment (jszakmeister) [inline modules/gradle-clojure-tools/build.gradle]: I didn't know about shimdandy before... it's pretty slick.  And a very nice addition to the plugin.

**Comment by ajoberstar** [comment]
Thanks for the review and sorry for the delayed response! On the serialization UID, I don't have a very firm grasp of when those are or aren't needed. I can add one if you think that's the right call.

**Comment by jszakmeister** [comment]
You're welcome!  I'm not real sure it's needed... I guess I don't understand what Gradle or the plugin intends by supporting it.  I think if it's just a runtime thing to communicate between the the processes, it's probably fine as-is.  If it's meant to be persisted, then I think we need it.  I have done much on the serialization front recently (it was years ago the last time I looked at anything related to it), but the [docs](https://docs.oracle.com/javase/7/docs/api/java/io/Serializable.html) say:

> If a serializable class does not explicitly declare a serialVersionUID, then the serialization runtime will calculate a default serialVersionUID value for that class based on various aspects of the class, as described in the Java(TM) Object Serialization Specification. However, it is *strongly recommended* that all serializable classes explicitly declare serialVersionUID values, since the default serialVersionUID computation is highly sensitive to class details that may vary depending on compiler implementations, and can thus result in unexpected InvalidClassExceptions during deserialization. Therefore, to guarantee a consistent serialVersionUID value across different java compiler implementations, a serializable class must declare an explicit serialVersionUID value. It is also strongly advised that explicit serialVersionUID declarations use the private modifier where possible, since such declarations apply only to the immediately declaring class--serialVersionUID fields are not useful as inherited members. Array classes cannot declare an explicit serialVersionUID, so they always have the default computed value, but the requirement for matching serialVersionUID values is waived for array classes.

Do you have a sense of what it's for?  If it's IPC, then let's just leave it be--otherwise you need to be concerned about changing the serialVersionUID when you [change the class](http://docs.oracle.com/javase/6/docs/platform/serialization/spec/version.html#6678), which is the biggest headache I remember about it. :-)

**Comment by ajoberstar** [comment]
Thanks for the detail. I believe it's all for IPC, so I think we're good. I'll merge this in now.

**Pull request merged**